domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt

```diff
@@ -0,0 +1,291 @@
+import org.oppia.android.app.model.ExplorationCheckpoint
+import org.oppia.android.app.model.ExplorationCheckpointDatabase
+import org.oppia.android.app.model.ExplorationCheckpointDetails
+import org.oppia.android.app.model.ProfileId
+import org.oppia.android.data.persistence.PersistentCacheStore
+import org.oppia.android.domain.oppialogger.OppiaLogger
+import org.oppia.android.util.data.AsyncResult
+import org.oppia.android.util.data.DataProvider
+import org.oppia.android.util.data.DataProviders
+import org.oppia.android.util.data.DataProviders.Companion.transformAsync
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private const val CACHE_NAME = "exploration_checkpoint_database"
+private const val RETRIEVE_EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID =
+  "retrieve_exploration_checkpoint_provider_id"
+private const val RETRIEVE_OLDEST_CHECKPOINT_DETAILS_DATA_PROVIDER_ID =
+  "retrieve_oldest_checkpoint_details_provider_id"
+private const val RECORD_EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID =
+  "record_exploration_checkpoint_provider_id"
+private const val DELETE_EXPLORATION_CHECKPOINT_DATA_PROVIDER_ID =
+  "delete_exploration_checkpoint_provider_id"
+
+/**
+ * Controller for saving, retrieving, updating, and deleting exploration checkpoints.
+ */
+@Singleton
+class ExplorationCheckpointController @Inject constructor(
+  private val cacheStoreFactory: PersistentCacheStore.Factory,
+  private val dataProviders: DataProviders,
+  private val oppiaLogger: OppiaLogger,
+  @ExplorationStorageDatabaseSize private val explorationCheckpointDatabaseSizeLimit: Int
+) {
+
+  /** Different stages in which the exploration checkpoint database can exist. */
+  enum class ExplorationCheckpointDatabaseState {
+
+    /** checkpoint database has not exceeded the allocated size limit. */
+    CHECKPOINT_DATABASE_SIZE_LIMIT_NOT_EXCEEDED,
+
+    /** checkpoint database has exceeded the allocated size limit. */
+    CHECKPOINT_DATABASE_SIZE_LIMIT_EXCEEDED
+  }
+
+  /** Indicates that no checkpoint was found for the specified explorationId and profileId. */
+  class ExplorationCheckpointNotFoundException(message: String) : Exception(message)
+
+  /**
+   * These Statuses correspond to the exception and the checkpoint database states above
+   * such that if the deferred result contains
+   *
+   * CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_EXCEEDED, the
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_EXCEEDED] will be
+   * passed to a successful AsyncResult.
+   *
+   * CHECKPOINT_SAVED_DATABASE_SIZE_LIMIT_NOT_EXCEEDED,
+   * [ExplorationCheckpointDatabaseState.CHECKPOINT_DATABASE_SIZE_LIMIT_NOT_EXCEEDED] will be
+   * passed to a successful AsyncResult.
+   *
+   * CHECKPOINT_NOT_FOUND, the [ExplorationCheckpointNotFoundException] will be passed to a failed
+   * AsyncResult.
+   *
+   * SUCCESS corresponds to successful AsyncResult with value as null.
+   */
+  enum class ExplorationCheckpointActionStatus {
```

Comment:
Also keep only one enum instead of two.

Reply:
moved this enum to a separate class because this enum will be used from other classes

Reply:
Is this moved or not? I am bit confused.

Reply:
the enum `ExplorationCheckpointActionStatus` is now in a new file.